core/utils/wasm-builder/src/wasm_project.rs

```diff
@@ -0,0 +1,162 @@
+// Copyright 2019 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+use std::{fs, path::{Path, PathBuf}, borrow::ToOwned, process::Command};
+
+use toml::value::Table;
+
+/// The name for the wasm binary.
+const WASM_BINARY: &str = "wasm_binary";
+
+/// Creates the WASM project, compiles the WASM binary and compacts the WASM binary.
+///
+/// # Returns
+/// The path to the compact WASM binary.
+pub fn create_and_compile(cargo_manifest: &Path) -> PathBuf {
+	let project = create_project(cargo_manifest);
+
+	build_project(&project);
+	compact_wasm_file(&project);
+
+	project.join(format!("{}.compact.wasm", WASM_BINARY))
+}
+
+/// Find the `Cargo.lock` relative to the `OUT_DIR`.
+fn find_cargo_lock() -> PathBuf {
+	let mut out_dir = build_helper::out_dir();
+
+	while out_dir.pop() {
```

Comment:
This is not great, as the target directory could be totally unrelated to where the `Cargo.lock` is (if you do `cargo b --target-dir=/tmp/foo` for example).

I don't think there's a way to detect that, except by depending on the `cargo` crate of crates.io, but that's a pretty big dependency to pull in the tree.

Reply:
Hmm :( Then the only solution I see, is to make people specify the correct dependency version in their `Cargo.toml` and I don't copy the `Cargo.lock`.

Reply:
> This is not great, as the target directory could be totally unrelated to where the `Cargo.lock` is (if you do `cargo b --target-dir=/tmp/foo` for example).
> 
> I don't think there's a way to detect that, except by depending on the `cargo` crate of crates.io, but that's a pretty big dependency to pull in the tree.

Could we implement a subset of Cargo’s logic ourselves, or get the info we need from environment variables?